### DISCUSSION TITLE
Automatic conversion of C# types in script

### DISCUSSION THREAD

**Discussion by ZmorzynskiK** [sebastienros/jint#1519] [Q&A] [answered]
I have the following question/issue when updating JINT (and scripts) from .NET 4.8 to .NET 6.

I'm using `MongoDB` types in js script (https://mongodb.github.io/mongo-csharp-driver/2.19/apidocs/html/T_MongoDB_Bson_BsonValue.htm).
Earlier when working (getting, creating) `BsonValue` in js, it was treated as .net type, so this was the code to get actual value:
```
var m = MongoDB.Bson.BsonValue.Create(2);
// m was BsonValue, so it required a function from BsonValue:
m.ToInt32();
```
But currently it looks like the `m` is already converted to js value (to `number` in this case).
I have a lot of scripts that use the `ToXXX()` pattern and I'm wondering how to handle this transition automatically, without the need to rewrite all the "old" scripts working with the older version?
Any ideas will be appreciated :)

thanks
Krzysztof

**Comment by lahma** [comment, marked as answer]
So I guess we are talking about 3.x beta? From and to version would help here.

So the ask here is that you would like to configure Jint not to do automatic number conversion for given type (even when possible) to keep your script's old behavior. 

Jint has become too smart for you 😉

**Reply by ZmorzynskiK** [reply]
> Jint has become too smart for you 😉

yes, looks that way :)

For old .NET 4.8 I use 3.0.0-beta-1715 version.
For new .NET 6 I use 3.0.0-beta-2048

Seems I just need to have this automatic conversion disabled for specific types, like `BsonValue`.
Any recommendations on how to do it?

**Reply by lahma** [reply]
You should be able achieve this via custom object converter:

```c#
[Fact]
public void CanKeepOriginalTypeForBsonTypes()
{
    var m = BsonValue.Create(2);
    var engine = new Engine(optiosn =>
    {
        optiosn.AddObjectConverter<BsonPreservingConverter>();
    });
    engine.SetValue("m", m);
    engine.SetValue("func", new Action<JsValue>((param1) =>
    {
        Assert.IsType<ObjectWrapper>(param1);
    }));

    engine.Evaluate("func(m)");
    Assert.Equal(2, engine.Evaluate("m.ToInt32();"));
}


public sealed class BsonPreservingConverter : IObjectConverter
{
    public bool TryConvert(Engine engine, object value, out JsValue result)
    {
        if (value is BsonValue)
        {
            result = new ObjectWrapper(engine, value);
            return true;
        }

        result = null;
        return false;
    }
}
```

**Reply by ZmorzynskiK** [reply]
Wow, thank you 🙏, works like a charm, this will definitely make my life easier 😉
Love Jint, it's a great way to make super-flexible and extensible C# applications.
Marking as answer.

all the best
Krzysztof
